### ISSUE TITLE
Add Arch Linux packaging support

### ISSUE THREAD

**Issue by gtaylor** [jordansissel/fpm#231]
I'd settle for just the PKGBUILD file generation, but generation of the package archives would be trivial to add if we could generate PKGBUILD files.

**Comment by djhaskin987** [comment]
Just a status update. all my work for this ticket is on the branch `feature/arch-support` on my forked version of this repository. The work is nearly complete; I just need to implement an arch specific option, add some more unit tests, fix the ones I've added, and test my work generally. Almost there!

**Comment by djhaskin987** [comment]
@hatt @jordansissel and anyone else who wants to help, I could use an extra pair of eyes.
My feature branch is so *very* near complete. I just hit this one last snag:

I wrote a test to make sure that permissions are properly preserved from the staging path's files:

```
Failures:

  1) FPM::Package::Pacman#output file permissions should preserve file permissions for /usr
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  2) FPM::Package::Pacman#output file permissions should preserve file permissions for /usr/bin
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  3) FPM::Package::Pacman#output file permissions should preserve file permissions for /usr/lib
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  4) FPM::Package::Pacman#output file permissions should preserve file permissions for /usr/share
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  5) FPM::Package::Pacman#output file permissions should preserve file permissions for /usr/share/doc
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  6) FPM::Package::Pacman#output file permissions should preserve file permissions for /etc
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  7) FPM::Package::Pacman#output file permissions should preserve file permissions for /var
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  8) FPM::Package::Pacman#output file permissions should preserve file permissions for /var/lib
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'

  9) FPM::Package::Pacman#output file permissions should preserve file permissions for /var/lib/foo
     Failure/Error: dir)).mode & 07777 } == perm
     Insist::Failure:
       Expected 493, but got 509
     # ./spec/fpm/package/pacman_spec.rb:164:in `block (5 levels) in <top (required)>'
```

Failure to properly preserve permissions makes arch linux sad:

resolving dependencies...
looking for inter-conflicts...

Packages (1): fu-:1.2.3-1

Total Installed Size:   0.06 MiB
Net Upgrade Size:       0.04 MiB

```
:: Proceed with installation? [Y/n] y
(1/1) checking keys in keyring                                            [#########################################] 100%
(1/1) checking package integrity                                          [#########################################] 100%
(1/1) loading package files                                               [#########################################] 100%
(1/1) checking for file conflicts                                         [#########################################] 100%
(1/1) checking available disk space                                       [#########################################] 100%
(1/1) downgrading fu                                                      [#########################################] 100%
warning: directory permissions differ on /etc/
filesystem: 755  package: 775
warning: directory permissions differ on /etc/profile.d/
filesystem: 755  package: 775
warning: directory permissions differ on /usr/
filesystem: 755  package: 775
warning: directory permissions differ on /usr/share/
filesystem: 755  package: 775
warning: directory permissions differ on /usr/share/doc/
filesystem: 755  package: 775
warning: directory permissions differ on /usr/lib/
filesystem: 755  package: 775
warning: directory permissions differ on /usr/bin/
filesystem: 755  package: 775
warning: directory permissions differ on /var/
filesystem: 755  package: 775
warning: directory permissions differ on /var/lib/
filesystem: 755  package: 775
```

Steps to reproduce:

1. Make a tarball with correct permissions of files: usr/ => 755, usr/bin => 755, usr/share/ => 755, etc.
2. Use the tarball to make a pacman package: `fpm -s tar -t pacman -n fu -v 1.2.3 fu.tar.gz`
3. Your resulting pacman package was written out with the "group write" permission enabled where it otherwise was not enabled in the tarball.
4. ARGH!

The functionality (and bug) is here: https://github.com/djhaskin987/fpm/tree/feature/arch-support . help? I don't know what to do.

**Comment by jordansissel** [comment]
493 is 0755, 509 is 0775. This is *usually* due to your environment's umask, but I don't have data to support that hypothesis yet.

**Comment by djhaskin987** [comment]
Fixed the problem. I needed to copy the metadata as well as the files from the `staging_path` to the build path. It's all ready to go! It both passes rspec and works with manual installation.

**Comment by djhaskin987** [comment]
![So excited](http://www.allgifs.com/wp-content/uploads/2013/08/excited.gif)

**Comment by hatt** [comment]
Good to hear, I was playing with your brand on my side and couldn't replicate the issue :disappointed:

Nice work @djhaskin987! I got held up on the mtree stuff and couldn't find proof anywhere that the size was effectively ignored, then life took over.

**Comment by djhaskin987** [comment]
Thanks for looking at that for me, I'm glad I'm not alone on this.

**Comment by flaccid** [comment]
Thanks for the hard work here, I'm really looking forward to this!

**Comment by djhaskin987** [comment]
This is, in my opinion, ready to go now. I just added compression option support.

**Comment by Anachron** [comment]
So what's missing here? Anything?

**Comment by djhaskin987** [comment]
@jordansissel is there anything else you would like me to add :cake: ?  Do you have any concerns :worried: :construction_worker: ?

**Comment by jordansissel** [comment]
I love :cake:

Will review!

**Comment by djhaskin987** [comment]
Thanks! I saw your review stuff, I'll take a look at it tonight :)

Daniel Jay Haskin
http://djhaskin987.github.io/

**Comment by infertux** [comment]
Is Arch Linux supposed to be supported as a target? I'm getting this error with fpm 1.4.0:

```bash
$ fpm -s gem -t pacman blah
Invalid output package (-t flag) type "pacman". Expected one of: cpan, deb, dir, empty, gem, npm, osxpkg, p5p, pear, pkgin, puppet, python, rpm, sh, solaris, tar, virtualenv, zip {:level=>:warn}
```

**Comment by djhaskin987** [comment]
It is supported on the development branch but it hasn't been released yet.

On March 19, 2016 3:29:50 PM MDT, "Cédric Félizard" <notifications@github.com> wrote:
>Is Arch Linux supposed to be supported as a target? I'm getting this
>error with fpm 1.4.0:
>
>```bash
>$ fpm -s gem -t pacman blah
>Invalid output package (-t flag) type "pacman". Expected one of: cpan,
>deb, dir, empty, gem, npm, osxpkg, p5p, pear, pkgin, puppet, python,
>rpm, sh, solaris, tar, virtualenv, zip {:level=>:warn}
>```
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jordansissel/fpm/issues/231#issuecomment-198787536

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by hatt** [comment]
Maybe we need a tag on PR or issues for `awaiting release`, since it seems like quite a few at the moment.

**Comment by Anachron** [comment]
I agree with @hatt :+1:

**Comment by jordansissel** [comment]
Good idea! Normally, there wouldn't be such delays between releases, but
master has several failing tests which, as a personal decision, have me
waiting until that is resolved before releasing.

❤️

**Comment by infertux** [comment]
Thanks for the clarification guys, I can confirm `-t pacman` works when building the Gem from the `master` branch.

(Agreed you shouldn't make a release with failing tests ;))

**Comment by hatt** [comment]
I had a quick look into the failing tests, it seems to be because the tar command is hardcoded. The following diff makes tests pass. What worries me is that some stuff relies specifically on gnutar while others on bsdtar. On OSX, only gnutar works for the packaging, but only bsdtar works for generating the mtree. If we're going to force bsdtar, then maybe we should make it get used for everything? Diff incoming:

```
diff --git a/lib/fpm/package/pacman.rb b/lib/fpm/package/pacman.rb
index eebc9eb..40b715c 100644
--- a/lib/fpm/package/pacman.rb
+++ b/lib/fpm/package/pacman.rb
@@ -96,7 +96,7 @@ class FPM::Package::Pacman < FPM::Package
   def input(pacman_pkg_path)
     control = {}
     # Unpack the control tarball
-    safesystem("tar", "-C", staging_path, "-xf", pacman_pkg_path)
+    safesystem(tar_cmd, "-C", staging_path, "-xf", pacman_pkg_path)
     pkginfo = staging_path(".PKGINFO")
     mtree = staging_path(".MTREE")
     install = staging_path(".INSTALL")
@@ -264,7 +264,7 @@ class FPM::Package::Pacman < FPM::Package

     File.expand_path(output_path).tap do |path|
       ::Dir.chdir(build_path) do
-        safesystem(*(["tar",
+        safesystem(*([tar_cmd,
                       compression_option,
                       "-cf",
                       path] + data_tar_flags + \
```

**Comment by jordansissel** [comment]
fpm 1.5.0 shipped with pacman support via #916.
